docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -0,0 +1,1438 @@
+  namespace: tenant-alpha
+spec:
+  trainingRuntimeRef:
+    name: torch-tune-llama-7b
+  datasetConfig:
+    storageUri: s3://dataset/custom-dataset/yelp-review
+    parameters:
+      split: train[:5000]
+  modelConfig:
+    input:
+      storageUri: hf://meta-llama/Llama-2-7b
+    output:
+      storageUri: s3://trained-model
+```
+
+### The Trainer API
+
+The `Trainer` represents the APIs that data scientists can use to configure the trainer settings.
+This trainer is executed on every distributed training Node.
+
+```golang
+type Trainer struct {
+
+  // Docker image for the Trainer.
+  Image string `json:"image,omitempty"`
+
+  // Command for the training container.
+  // Validate that command contains torchrun or mpirun.
+  Command []string `json:"command,omitempty"`
+
+  // Args for the training container.
+  Args []string `json:"args,omitempty"`
+
+  // Env for the training container.
+  Env []corev1.EnvVar `json:"env,omitempty"`
+
+  // Number of training nodes.
+  NumNodes *int32 `json:"numNodes,omitempty"`
+
+  // Resource for each node.
+  ResourcesPerNode []corev1.resources `json:"resourcesPerNode,omitempty"`
+
+  // Number of processes in a single node.
+  // By default this value == number of GPUs in resources limits.
+  NumProcPerNode *int32 `json:"numProcPerNode,omitempty"`
+}
+```
+
+The following table explains how `TrainingRuntime` parameters will be overridden with `Trainer`.
+
+<table>
+  <tr>
+   <td><strong><code>Trainer</code> Parameter</strong>
+   </td>
+   <td><strong> <code>TrainingRuntime</code> Parameter</strong>
+   </td>
+  </tr>
+  <tr>
+   <td><code>.image</code>
+   </td>
+   <td><code>.spec.replicatedJobs[name=’Node’].template.spec.template.spec.containers[name=’trainer’].image</code>
+   </td>
+  </tr>
+  <tr>
+   <td><code>.command</code>
+   </td>
+   <td><code>.spec.replicatedJobs[name=’Node’].template.spec.template.spec.containers[name=’trainer’].command</code>
+   </td>
+  </tr>
+  <tr>
+   <td><code>.args</code>
+   </td>
+   <td><code>.spec.replicatedJobs[name=’Node’].template.spec.template.spec.containers[name=’trainer’].args</code>
+   </td>
+  </tr>
+  <tr>
+   <td><code>.env</code>
+   </td>
+   <td><code>.spec.replicatedJobs[name=’Node’].template.spec.template.spec.containers[name=’trainer’].env</code>
+   </td>
+  </tr>
+  <tr>
+   <td><code>.numNodes</code>
+   </td>
+   <td><code>.spec.numNodes</code>
+   </td>
+  </tr>
+  <tr>
+   <td><code>.resourcesPerNode</code>
+   </td>
+   <td><code>.spec.replicatedJobs[name=’Node’].template.spec.template.spec.containers[name=’trainer’].resources</code>
+   </td>
+  </tr>
+</table>
+
+### The Dataset Config API
+
+The `DatasetConfig` represents the APIs that data scientists can use to configure the dataset location.
+
+```golang
+type DatasetConfig struct {
+
+	// Storage uri for the dataset provider.
+	StorageUri string `json:"storageUri"`
+
+	// Custom parameters for the dataset initializer.
+	Parameters *[string]string `json:"parameters,omitempty"`
```

Comment:
```suggestion
	Parameters map[string]string `json:"parameters,omitempty"`
```

Reply:
I still wondering if we can have scheme safe union type parameters. Because this map parameters do not allow us to add any CRD based validation and so on. I'm suspecting that this non scheme safe parameters API could be blocker for introducing the CRD validations.

But, I'm ok with this non scheme safe parameters for the alpha API. Before we graduate API to Beta, we can revisit here, and we can consider which parameters we should define as dedicated field.

Reply:
Basically, I agree with parameter fields in alpha API since it's challenging to decide which model/storage provider parameters should be. defined in the API during the first implementation. So, I would like to revisit here after first implementation and collected feedback from users.

Reply:
Sounds good!